### PR TITLE
[shelly]Enhancements and bug fixes

### PR DESCRIPTION
**Pull request by markus7017**
This PR includes all changes from #6592 and replaces it. PR #6592 has been closed.

Signed-off-by: markus7017 <markus7017@gmail.com>

### CODE REVIEW

**Comment by markus7017** [comment]
PR #6728 will solve most of the Travis warnings

**Comment by markus7017** [comment]
@kaikreuzer Who will be the 2nd reviewer?

**Comment by markus7017** [comment]
I need to put the PR on-hold. It seems that there is an inconsistency in the repo. I saw that some changes are missing, which I already made in 2.5.0

**Comment by markus7017** [comment]
@kaikreuzer When could I expect the review? What's the timeline for 2.5.2?
I expect shipping of new device types early next week. Would be good if we could merge the changes until then so I could open a new PR.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 1]
```diff
@@ -116,7 +118,6 @@ public ShellyBaseHandler(Thing thing, ShellyBindingConfiguration bindingConfig,
      */
     @Override
     public void initialize() {
-        updateStatus(ThingStatus.UNKNOWN);
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: Why did you remove this line?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 2]
```diff
@@ -228,13 +235,14 @@ private void initializeThing() throws IOException {
         profile = tmpPrf; // all initialization done, so keep the profile
 
         // Validate device mode
-        String thingType = getThing().getThingTypeUID().getId();
         String reqMode = thingType.contains("-") ? StringUtils.substringAfter(thingType, "-") : "";
         if (!reqMode.isEmpty() && !tmpPrf.mode.equals(reqMode)) {
+            logger.info(
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: Why did you remove the message from the Thing status? The Thing status is the correct place for it.

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml [hunk 3]
```diff
@@ -39,8 +45,71 @@
...
+			<label>Enable CoIoT Events</label>
+			<description>If enabled the Coap protocol will be used to receive status updates.</description>
+			<advanced>true</advanced>
+			<default>false</default>
+		</parameter>
+		<parameter name="updateInterval" type="integer" required="true">
+			<label>Update Interval</label>
+			<description>Interval in seconds to query an update from the device.</description>
+			<default>60</default>
+			<unitLabel>sec</unitLabel>
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]:
```suggestion
			<unitLabel>s</unitLabel>
```

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml [hunk 4]
```diff
@@ -39,8 +45,71 @@
 			<label>Enable Push Events (short/long)</label>
 			<description>True if the binding should register to get the Short/Long Push Events.</description>
 			<advanced>true</advanced>
+			<default>false</default>
+		</parameter>
+		<parameter name="eventsCoIoT" type="boolean" required="false">
+			<label>Enable CoIoT Events</label>
+			<description>If enabled the Coap protocol will be used to receive status updates.</description>
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]:
```suggestion
			<description>If enabled, the CoAP protocol is used to receive status updates.</description>
```

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml [hunk 5]
```diff
@@ -23,6 +23,12 @@
 			<description>IP-Address of the Shelly device.</description>
 			<context>network-address</context>
 		</parameter>
+		<parameter name="weakSignal" type="integer" required="false">
+			<label>Weak Signal Threshold</label>
+			<description>An alarm is triggered if a device reports WiFi signal strength below this threshold. Default: -80 dBm</description>
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: This feature does imho not make much sense in a binding - it is not at all specific here.
If a user wishes such an alarm, wouldn't it be hundred times better to have the alarm cover ALL devices that report their signal strength and not just the Shelly ones? This hence belongs into a rule and not in a binding.

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml [hunk 6]
```diff
@@ -39,8 +45,71 @@
...
+			<label>Password</label>
+			<description>Password for HTTP API access.</description>
+			<context>password</context>
+		</parameter>
+		<parameter name="deviceIp" type="text" required="true">
+			<label>Device IP Address</label>
+			<description>IP-Address of the Shelly device.</description>
+			<context>network-address</context>
+		</parameter>
+		<parameter name="weakSignal" type="integer" required="false">
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: see above

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml [hunk 7]
```diff
@@ -89,11 +158,19 @@
 			<advanced>true</advanced>
 			<default>true</default>
 		</parameter>
+		<parameter name="weakSignal" type="integer" required="false">
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: see above

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml [hunk 8]
```diff
@@ -365,4 +353,20 @@
 			</options>
 		</event>
 	</channel-type>
+
+	<channel-type id="timeoutErrors" advanced="true">
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml]: No, please don't add such debugging channels to the Things, this really feels not right.
Doing debug level logging should normally be enough. If you really want it somewhere else, maybe it would be ok to add it as Thing properties.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 1 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: I keep the status INITIALIZING until the device is fully initialized. For non-battery devices this will be done with the next status update, for battery-powered devices this could take longer until the device wakes up. UNKOWN sounds like an error, even I now that the device is not yet initialized

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 5 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: ok, but then it should be a OH core feature. The binding can't monitor the wifi signal for other devices. Same argument: Wha does each user has to add a rule for each device (or one serving it all). Would be a cool feature for OH3 to extend the channel type with a health monitoring.

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 6 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: same :-)

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: dto

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml]: ok, was designed as a temp at the beginning, but provided some good feedback
nevertheless I'll move it to thing properties as stats values. I'm currently working with the community and the Shelly guys on REST timeouts. Various users are reporting issues whereas others (like me) see timeouts very rarely. Once the problem is fixed (maybe a firmware issue) I Could remove those counters anyways.

**Comment by markus7017** [comment]
> Thanks, sorry to keep you waiting, but I have very busy times atm...

No problem. If would help me to provide a stable release to the community, but also, because I have issues with my repo and want to re-initialize the repo after this PR has merged.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: My understanding is that
updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                    "@text/offline.conf-error-wrong-mode")
pulls up the language resource "offline.conf-error-wrong-mode" to get this messages from shelly_de.properties/shelly.properties. Isn't it this the case?

So a info message goes to the log and the thing status gets updated with the genral message - or is there a way to pass parameters to the messages retrived from the properties file?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: I am pretty sure that we discussed this before (when adding that line), but I can certainly again point you to the developer docs: https://github.com/openhab/openhab-core/blob/master/tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java#L69-L79

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 2 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: > is there a way to pass parameters to the messages retrived from the properties file?

There is, see https://www.openhab.org/docs/developer/utils/i18n.html#i18n-text-provider-api.

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 5 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: Yes, the signal strength has been defined as a system channel to allow such feature for the core. The new UI should definitely show the signal strength for things and we could also add further functionality for that.

Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: ok, what do you think of
- keep the feature in 2.5
- remove it for 3.0 and integrate such monitoring in core

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: changed

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 4 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: changed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 1 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: ok, will be changed

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml]: moved to thing properties

**Comment by markus7017** [comment]
@kaikreuzer The above things are implemented (not yet checked in), any additional input?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: I don't see any change?

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 5 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: Please no. Stuff that is there usually stays around far too long. And if it is even something that is a workaround for a missing feature, it will be copied by others and it'll be all over the place.

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 4 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: did you maybe miss to push your changes?

**Comment by kaikreuzer** [comment]
@markus7017 I cannot see any recent commit from you, I guess you still need to push it :-)

**Comment by markus7017** [comment]
> @markus7017 I cannot see any recent commit from you, I guess you still need to push it :-)

@kaikreuzer It seems that PR is based on the wrong source branch. It shows 2.5.x also as source branch in my repo, which I used to adapt to 2.5.x, then I switched to shelly-snapshot-2.5.x. Obviously I used the wrong branch to create this PR. Could we fix this or do I need to re-create the PR? Latest status was committed to shelly-snapshot-2.5.x

**Comment by kaikreuzer** [comment]
Well, just delete your local 2.5.x branch, rename shelly-snapshot-2.5.x to 2.5.x and do a force push. That should bring this PR to the code revision you have on your shelly-snapshot-2.5.x branch.

**Comment by kaikreuzer** [comment]
Alternatively do a cherry pick of the new commits from shelly-snapshot-2.5.x to 2.5.x and do a normal push.

**Comment by markus7017** [comment]
@kaikreuzer Ok, should be ready to continue. Last change is setting status UNKNOWN on initialization and this shows up here and in my 2.5.x branch.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 9]
```diff
@@ -631,6 +683,39 @@ public boolean requestUpdates(int requestCount, boolean refreshSettings) {
...
 
+    /**
+     * Map input states to channels
+     *
+     * @param groupName Channel Group (relay / relay1...)
+     *
+     * @param status Shelly device status
+     * @return true: one or more inputs were updated
+     */
+    @SuppressWarnings({ "null", "unused" })
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: Why do you have to suppress unused? Can't you simply remove unused variables?
Also, please try to avoid null suppressions and rather refactor the code to be fully null-safe.
Ok, if you address this in a future PR, no blocker wrt merging now.

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 5 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: This is still here...

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 6 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: Also still here...

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 7 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: dito

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml (hunk 8 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml]: still here - more commits missing?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 1 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: done

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 5 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: removed even I don't agree

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 6 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: dto

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/config/config.xml]: dito

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 9 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: because Eclipse sometimes just report BS

**Comment by markus7017** [comment]
@kaikreuzer done

File: bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/channels.xml]: left over, removed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 9 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: example: For this statement Eclipse reports a redundant null check

`
Strong thingName = properties.get(PROPERTY_SERVICE_NAME) != null ? properties.get(PROPERTY_SERVICE_NAME).toLowerCase()
                : thingType;
`
how to resolved those warnings?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: ok, I figured out the i18nProvider stuff, but how to format the message in the properties file to inject parameters?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 9 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: Which is correct, because theoretically another thread could meanwhile remove the property before you call the get.
This should solve those issues:
```
String thingName = properties.get(PROPERTY_SERVICE_NAME);
thingName = thingName!=null ? thingName.toLowerCase() : thingType;
```

Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: nope, not it's "thingName!=null"

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: @kaikreuzer could you point me to the solution?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 9 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: Then probably this way:
```
```
String serviceName = properties.get(PROPERTY_SERVICE_NAME);
String thingName = serviceName!=null ? serviceName.toLowerCase() : thingType;
```
That's anyhow cleaner than my previous suggestion.
